File: csharp/solution/Webshop.UnitTests/UserServiceTests.cs
using NUnit.Framework;
using Webshop.Service.Common;
using Webshop.Service.Users;

namespace Webshop.UnitTests;

[TestFixture]
public class UserServiceTests
{
    [Test]
    public void CreateUser_WhenPasswordIsInvalid_ShouldThrowException()
    {
        var userService = new UserService(new UserRepository(), new PasswordValidator());

        // Alternative 1 (manually checking that exception is thrown):
        bool exceptionThrown = false;
        try
        {
            userService.CreateUser("valid@email.com", "Test", "User", "12345678");
        }
        catch (InvalidPasswordException)
        {
            exceptionThrown = true;
        }
        Assert.IsTrue(exceptionThrown);

        // Alternative 2 (using Assert.Throws):
        //Assert.Throws<InvalidPasswordException>(() => userService.CreateUser("valid@email.com", "Test", "User", "12345678"));
    }

    [Test]
    public void CreateUser_WhenPasswordIsValid_ShouldCreateNewUser()
    {
        var userService = new UserService(new UserRepository(), new PasswordValidator());
        var user = userService.CreateUser("valid@email.com", "Test", "User", "password123");

        Assert.That(user, Is.Not.Null);
    }

    [Test]
    public void CreateUser_WhenUserIsCreated_ShouldBeAssignedAnId()
    {
        var userService = new UserService(new UserRepository(), new PasswordValidator());
        var user = userService.CreateUser("valid@email.com", "Test", "User", "password123");

        Assert.That(user.Id, Is.Not.Null.Or.Empty);
    }

    [Test]
    public void CreateUser_WhenUserIsCreated_ShouldBeRetrievable()
    {
        var userService = new UserService(new UserRepository(), new PasswordValidator());

        var user = userService.CreateUser("valid@email.com", "Test", "User", "password123");
        var userFromDb = userService.GetUser(user.Id);

        Assert.That(userFromDb, Is.Not.Null);
    }
}

File: java/solution/src/test/java/Webshop/UnitTests/UserServiceTests.java
package Webshop.UnitTests;

import org.junit.jupiter.api.Test;
import Webshop.Service.Users.InvalidPasswordException;
import Webshop.Service.Common.PasswordValidator;
import Webshop.Service.Users.UserRepository;
import Webshop.Service.Users.User;
import Webshop.Service.Users.UserService;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

public class UserServiceTests {
    @Test
    public void createUser_whenPasswordIsInvalid_shouldThrowException() {
        UserService userService = new UserService(mock(UserRepository.class), new PasswordValidator());

        assertThrows(InvalidPasswordException.class, () -> userService.createUser("user1", "valid@email.com", "Test", "User", "12345678"));
    }

    @Test
    public void createUser_whenPasswordIsValid_shouldCreateNewUser() throws InvalidPasswordException {
        UserRepository userRepository = mock(UserRepository.class);
        UserService userService = new UserService(userRepository, new PasswordValidator());

        when(userRepository.addUser("user1","valid@email.com", "Test", "User", "password123")).thenReturn(new User("user1", "valid@email.com", "Test", "User", "password123"));

        User user = userService.createUser("user1","valid@email.com", "Test", "User", "password123");

        assertNotNull(user);
    }

    @Test
    public void createUser_whenUserIsCreated_shouldBeAssignedAnId() throws InvalidPasswordException {
        UserRepository userRepository = mock(UserRepository.class);
        UserService userService = new UserService(userRepository, new PasswordValidator());

        when(userRepository.addUser("user1","valid@email.com", "Test", "User", "password123")).thenReturn(new User("user1", "valid@email.com", "Test", "User", "password123"));

        User user = userService.createUser("user1","valid@email.com", "Test", "User", "password123");

        assertNotNull(user.getId());
    }

    @Test
    public void createUser_whenUserIsCreated_shouldBeRetrievable() throws InvalidPasswordException {
        UserRepository userRepository = mock(UserRepository.class);
        UserService userService = new UserService(userRepository, new PasswordValidator());

        when(userRepository.addUser("user1","valid@email.com", "Test", "User", "password123")).thenReturn(new User("user1", "valid@email.com", "Test", "User", "password123"));
        when(userRepository.getUser("user1")).thenReturn(new User("user1", "valid@email.com", "Test", "User", "password123"));

        User user = userService.createUser("user1","valid@email.com", "Test", "User", "password123");
        User userFromDb = userService.getUser(user.getId());

        assertNotNull(userFromDb);
    }
}

